### ISSUE TITLE
Possible solution for dumping, loading and caching registry.

### ISSUE THREAD

**Issue by hgrecco** [hgrecco/pint#1462]
I have found that with a minor change to `Context`

```diff
diff --git a/pint/context.py b/pint/context.py
index 59ea9cf..2a8376b 100644
--- a/pint/context.py
+++ b/pint/context.py
@@ -118,6 +118,15 @@ class Context:
         #: Used as a convenience dictionary to be composed by ContextChain
         self.relation_to_context = weakref.WeakValueDictionary()

+    def __getstate__(self):
+        attributes = self.__dict__.copy()
+        attributes['relation_to_context'] = dict(self.relation_to_context)
+        return attributes
+
+    def __setstate__(self, state):
+        state["relation_to_context"] = weakref.WeakValueDictionary(state["relation_to_context"])
+        self.__dict__ = state
+
     @classmethod
     def from_context(cls, context: Context, **defaults) -> Context:
         """Creates a new context that shares the funcs dictionary with the
```
This should fix #1262

but also, after this change I can use [Cloudpickle](https://github.com/cloudpipe/cloudpickle) to serialize a UnitRegistry!

This could be very useful for #522 and other applications. 

However, I am still hesitant to natively automatically cache the definition file and then unpickle it until we are sure that this does not have any undesired side effects. But maybe we can apply this change to `Context` to enable testing this?

What to you think?

**Comment by hgrecco** [comment]
This is an example of how this can be used to speed up registry loading in the future. 

```python

import os
import hashlib
import pathlib
import cloudpickle

import pint

def build_tree(filename, is_resource=False):
    yield filename, os.stat(filename).st_mtime

    # This logic is taken from the registry
    with open(filename, encoding="utf-8") as fp:
        for line in fp:
            if line.startswith("@import"):
                if is_resource:
                    path = line[7:].strip()
                else:
                    try:
                        # path = os.path.dirname(file.name)
                        path = os.path.dirname(filename)
                    except AttributeError:
                        path = os.getcwd()
                    path = os.path.join(path, os.path.normpath(line[7:].strip()))
                yield from build_tree(path)


def load_registry(p: pathlib.Path):
    with p.open(mode="rb") as fi:
        ureg = cloudpickle.load(fi)

    # This hack tells me that something is not completely right.
    # These classes are attached to the registry, created dinamically
    # an not properly saved.
    # Calling init class recreate them, but somehow reference to those
    # in the registry work fine.
    ureg._init_dynamic_classes()
    return ureg


def build_registry(filename):

    # Build a tuple with all files to be parsed and their modification time.
    files = tuple(build_tree(filename))
    print(f"files: {files}")
    hd = hashlib.sha1(cloudpickle.dumps(files)).hexdigest()

    p = pathlib.Path(".").joinpath(hd).with_suffix(".cloudpickle")

    print(f"Cache file: {p}")
    try:
        ureg = load_registry(p)
        print("Loaded registry from cache.")
        return ureg
    except FileNotFoundError:
        print("File not found.")
    except Exception as ex:
        print(f"Could not load cache: {ex}")

    print("Creating registry from definition files.")
    ureg = pint.UnitRegistry(filename)
    with p.open(mode="wb") as fo:
        cloudpickle.dump(ureg, fo)
    print(f"Registry was written to cache file.")
    return ureg

ureg = build_registry("pint/default_en.txt")

print(ureg.parse_expression("4 meter"))
print(4 * ureg.meter)
```

Yes, it stills loads all text files due to the useful but here inconvenient include directive. However, it is still faster (between 50x and 100x in my benchmarks)

**Comment by hgrecco** [comment]
We merged #1464, which I think is a better way to go.
